### PR TITLE
[Improvement#5083][API] Refactored data source creation

### LINKED ISSUE #5083: [Improvement][API] Refactored data source creation
**Describe the question**
If we add a new datasource type in the future, we need to make a lot of changes to the code.
For example in `DataSourceController.createDataSource()`. 
We may need to add parameters like `javaSecurityKrb5Conf` and add `required = false`, because other datasource doesn't need the parameter.(This make the method need to receive a large number of parameters and reduce readability).
And in `DataSourceServiceImpl.buildParameter`
We may need to make the same modification. But for any type of datasource, we do the same logic for adding operation.
```
1. check params
2. transform params
3. insert to mysql
```
We need not change the core code

**What are the current deficiencies and the benefits of improvement**
Use a target class to wrap parameters of datasource, and add subclass for each datasource, for example:
```
@JsonTypeInfo(use = JsonTypeInfo.Id.CLASS, property = "type")
@JsonSubTypes({
        @JsonSubTypes.Type(value = MysqlDatasourceParams.class, name = "1"),
        @JsonSubTypes.Type(value = OracleDatasourceParams.class, name = "2"),
})
public abstract class AbstractDataDourceParams {
    
    public abstract Integer getType();
}
```
The request from the front can be automatically bound to the subclass through jackson deserialization, and the specific data source can be known through the type.
And add check processor or transform processor for each datasource to transform the params to metadata stored in mysql. After that, we needn't change any code when add a new kind of datasource, we just add a new subclass of datasource and add a new transform processor.

**Which version of DolphinScheduler:**
 -[dev]
 -[1.3.6]

### PR DESCRIPTION
**Pull request by ruanwenjun**
* Refact create/ update datasource API
  
## Purpose of the pull request

Increasing code scalability, make it ease to add a new datasource type for developer. More detail can see #5083 

This also closes #5083

## Brief change log

Change the create/update datasource method in DataSourceController

## Verify this pull request

Add Unit test and can be tested from ui.

### CODE REVIEW

**Comment by CalvinKirs** [comment]
@zhuangchong please take a look.thx

**Comment by chengshiwen** [comment]
@ruanwenjun Hi, sorry for too late review, could you solve the `18 Code Smells` and `2.8% Duplication` at first? Thanks so much

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/utils/datasource/DatasourceParamUtil.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+     *
+     * @param baseDataSourceParamDTO datasource param
+     * @throws ServiceException if param invalid
+     */
+    public static void checkDatasourceParam(BaseDataSourceParamDTO baseDataSourceParamDTO) {
+        Integer datasourceType = baseDataSourceParamDTO.getType().getCode();
+        if (!checkProcessorMap.containsKey(datasourceType)) {
+            throw new ServiceException(Status.DATASOURCE_DB_TYPE_ILLEGAL);
+        }
+        checkProcessorMap.get(datasourceType).checkDatasourceParam(baseDataSourceParamDTO);
```
Comment (zhuangchong) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/utils/datasource/DatasourceParamUtil.java]: Use switch, no need to use map to get.

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/utils/datasource/AbstractDatasourceProcessor.java [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+import org.apache.dolphinscheduler.spi.utils.JSONUtils;
+import org.apache.dolphinscheduler.spi.utils.StringUtils;
+
+import org.apache.commons.collections4.MapUtils;
+
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.regex.Pattern;
+
+public abstract class AbstractDatasourceProcessor implements DatasourceProcessor {
```
Comment (zhuangchong) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/utils/datasource/AbstractDatasourceProcessor.java]: The relationship between AbstractDatasourceProcessor and BaseDataSourceParamDTO and BaseDataSource

AbstractDatasourceProcessor and BaseDataSourceParamDTO are used to produce database parameters,

BaseDataSource is used to parse the parameters, it is recommended to consider BaseDataSource to modify together.

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/utils/datasource/AbstractDatasourceProcessor.java [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+        }
+        StringBuilder otherSb = new StringBuilder();
+        for (Map.Entry<String, String> entry : map.entrySet()) {
+            otherSb.append(String.format("%s=%s%s", entry.getKey(), entry.getValue(), separator));
+        }
+        if (DbType.DB2.equals(dbType)) {
+            otherSb.deleteCharAt(otherSb.length() - 1);
+        }
+        return otherSb.toString();
+    }
```
Comment (zhuangchong) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/utils/datasource/AbstractDatasourceProcessor.java]: Please implement it in the subclass.

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/utils/datasource/DatasourceParamUtil.java (hunk 1 above)
Comment (ruanwenjun) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/utils/datasource/DatasourceParamUtil.java]: Ok

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/utils/datasource/AbstractDatasourceProcessor.java (hunk 2 above)
Comment (ruanwenjun) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/utils/datasource/AbstractDatasourceProcessor.java]: Very good suggestion

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/utils/datasource/AbstractDatasourceProcessor.java (hunk 3 above)
Comment (ruanwenjun) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/utils/datasource/AbstractDatasourceProcessor.java]: Sure

**Comment by ruanwenjun** [comment]
@zhuangchong Hi, Zhaungchong, I have re-modified according  to your suggestion, please help me review. And there seems some problem with code coverage calculate, some unit test are not added to the `SonarCloud`.

**Comment by zhuangchong** [comment]
> @zhuangchong Hi, Zhaungchong, I have re-modified according to your suggestion, please help me review. And there seems some problem with code coverage calculate, some unit test are not added to the `SonarCloud`.

ok, let me take a look, please solve code coverage and code smell.

**Review by zhuangchong — APPROVED** [review]
LGTM.

**Review by CalvinKirs — APPROVED** [review]
LGTM

**Pull request merged**